cmd/xl-storage-format-v2.go

```diff
@@ -0,0 +1,437 @@
+package cmd
+
+import (
+	"bytes"
+	"encoding/binary"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+const (
+	// xlMetaVersion specifies version 200
+	xlMetaVersion uint16 = 200
+
+	// xlFormat specifies format type
+	xlFormat uint16 = iota
+
+	// XL metadata file carries per object metadata.
+	xlStorageFormatFile = "xl.meta"
+)
+
+func isXLVersion2(buf []byte) bool {
+	if len(buf) <= 4 {
+		return false
+	}
+	if binary.LittleEndian.Uint16(buf[:2]) == xlMetaVersion {
+		if binary.LittleEndian.Uint16(buf[2:4]) == xlFormat {
+			return true
+		}
+	}
+	return false
+}
+
+// The []journal contains all the different versions of the object.
+//
+// This array can have 3 kinds of objects:
+//
+// ``object``: If the object is uploaded the usual way: putobject, multipart-put, copyobject
+//
+// ``link``: If the object was copied(restored) from the older version to the latest using the copyobject API.
+//           This an optimization so that when an older version object is restored we don't copy the data,
+//           but just create metadata in the xl.meta link and object will share the same backend data directory.
+//
+// ``delete``: This is the delete-marker
+
+// The last element of this array is the latest. We will support versioning on newly created buckets.
+// i.e we don't support versioning on existing objects.
+
+// backend directory tree structure:
+
+// disk1/
+// └── bucket
+//     └── object
+//         ├── a192c1d5-9bd5-41fd-9a90-ab10e165398d
+//         │   └── part.1
+//         ├── c06e0436-f813-447e-ae5e-f2564df9dfd4
+//         │   └── part.1
+//         ├── df433928-2dcf-47b1-a786-43efa0f6b424
+//         │   └── part.1
+//         └── xl.meta
+
+//go:generate msgp -file=$GOFILE -tests=false
+
+// JournalType defines the type of journal type of the current entry.
+type JournalType uint8
+
+// List of different types of journal type
+const (
+	ObjectType JournalType = 0
+	DeleteType JournalType = 1
```

Comment:
Reserve `0` as an invalid value, so the zero value isn't accidentally picked up.

```suggestion
	invalidJournalType JournalType = 0
	ObjectType         JournalType = 1
	DeleteType         JournalType = 2
```